dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpServiceContextImpl.java

```diff
@@ -2288,62 +2290,69 @@ final protected void sendHeaders(HttpBaseMessageImpl msg) throws IOException {
-
-            ((NettyResponseMessage) getResponse()).processCookies();
-            HeaderHandler headerHandler = new HeaderHandler(myChannelConfig, response);
-            headerHandler.complianceCheck();
-            String closeNonUpgraded = (String) (this.myVC.getStateMap().get(TransportConstants.CLOSE_NON_UPGRADED_STREAMS));
-            // Shouldn't close upgraded requests
-            boolean upgradedRequest = closeNonUpgraded != null && closeNonUpgraded.equalsIgnoreCase("true");
-            if (!upgradedRequest && (!myChannelConfig.isKeepAliveEnabled() || (myChannelConfig.getMaximumPersistentRequests() != -1 && nettyContext.channel().attr(NettyHttpConstants.NUMBER_OF_HTTP_REQUESTS).get() >= myChannelConfig.getMaximumPersistentRequests()))) {
-                // Keep alive disabled or exceeded maximum number of keep alive requests
-                if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                    Tr.debug(tc, "sendHeaders: Adding close connection header due to keep alive disabled or exceeded number of maximum persistent requests");
+        // check compression and set up the Content-Encoding header if need be
+        if (null != this.compressHandler) {
+            ContentEncodingValues ce = this.compressHandler.getContentEncoding();
+            getResponse().setContentEncoding(ce);
+        }else {
+            String acceptEncoding = nettyContext.channel().attr(NettyHttpConstants.ACCEPT_ENCODING).get();
+            acceptEncoding = nettyRequest.headers().get(HttpHeaderKeys.HDR_ACCEPT_ENCODING.getName());
+            if (acceptEncoding != null) {
+                ResponseCompressionHandler compressionHandler = new ResponseCompressionHandler(getHttpConfig(), nettyResponse, acceptEncoding);
+                compressionHandler.setCurrentContentLength(getResponse().getContentLength());
+                compressionHandler.process();
+                if (compressionHandler.getEncoding() != null) {
+                    setupCompressionHandler(compressionHandler.getEncoding());
+                    ContentEncodingValues ce = this.compressHandler.getContentEncoding();
+                    getResponse().setContentEncoding(ce);
                 }
-                response.headers().set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
             }
-            if (HttpUtil.isContentLengthSet(response)) {
-                this.nettyContext.channel().attr(NettyHttpConstants.CONTENT_LENGTH).set(HttpUtil.getContentLength(response));
+        }
+        ((NettyResponseMessage) getResponse()).processCookies();
+        HeaderHandler headerHandler = new HeaderHandler(myChannelConfig, response);
+        headerHandler.complianceCheck();
+        String closeNonUpgraded = (String) (this.myVC.getStateMap().get(TransportConstants.CLOSE_NON_UPGRADED_STREAMS));
+        // Shouldn't close upgraded requests
+        boolean upgradedRequest = closeNonUpgraded != null && closeNonUpgraded.equalsIgnoreCase("true");
```

Comment:
boolean upgradedRequest = "true".equalsIgnoreCase(closeNonUpgraded)

Reply:
Created https://github.com/OpenLiberty/open-liberty/issues/32231 to follow up